### ISSUE TITLE
include metadata spreadsheet in staging area

### ISSUE THREAD

**Issue by amnonkhen** [ebi-ait/dcp-ingest-central#845]
We would like to include the metadata spreadsheet in the staging area so it can be made available for consumers in the Data Browser.

See [mockups and references & directory structure](https://miro.com/app/board/uXjVPbKHNuc=/) and [Slack channel](https://humancellatlas.slack.com/archives/C03VB103H63)

Tasks for ingest team:
- [x] add `DCP/2 Ingest` as source HumanCellAtlas/metadata-schema#1490
- [x] new EDAM ontology term ([ticket])(https://github.com/edamontology/edamontology/issues/850)
- include metadata spreadsheet with files in Staging Area
  - [x] generate the spreadsheet as a "fake" supplementary file - just in the staging area, but not in ingest db.
  - [x] testing
  - [x] Deploy to prod:
    - [x] core
    - [x] exporter
    - [x] broker 
- [x] connect spreadsheet generation to the Wrangling process

Tasks for data browser, ux teams:
- [ ] agree on appearance in the browser

Software Issues:
- #873

Further Tickets:
- [ ] fix spreadsheet format
- [ ] generate spreadsheets for everything in the browser
- [ ] generate spreadsheets for everything in the catalogue

**Comment by ESapenaVentura** [comment]
@amnonkhen and @Mil-m made a design chart - to be linked. 

Continuing today

**Comment by ESapenaVentura** [comment]
First 2 items done - We're going to go with [database entry metadata](https://ontology.archive.data.humancellatlas.org/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_2193) for the ontology term

**Comment by ESapenaVentura** [comment]
We can start testing this in staging

**Comment by ESapenaVentura** [comment]
@Mil-m and @amnonkhen working on this

Working on it today

**Comment by MightyAx** [comment]
Ingest Broker cannot be built until we we resolve the dependancy issues.

Hopefully by merging these PRs:
- ebi-ait/ingest-client#58 Deployed
- ebi-ait/geo_to_hca#41 Deployed

**Comment by Mil-m** [comment]
Launching the tests for ingest-broker locally

**Comment by ofanobilbao** [comment]
We are approaching the end of this task. Alexie helped Milena yesterday and @amnonkhen will work with her today.

**Comment by gabsie** [comment]
Continue with the tests today. @Mil-m

**Comment by idazucchi** [comment]
two things to fix: 

1. ~Currently every time someone hits the generate button a new file record will be added to the submission, but actually we only ever want one of these records within a submission, so if an old file exists we should update the existing file.~
1. ~We should add a wait after adding a file before sending the s3 command until the file is metadata invalid, otherwise the upload service will try to update the file to add the cloud url and will be denied due to the file being in the validating state.~

No longer relevant as we moved to "fake supplementary file"

**Comment by MightyAx** [comment]
Point 2 has been provisionally resolved:
https://github.com/ebi-ait/ingest-core/pull/146

**Comment by MightyAx** [comment]
3. We should also patch the project **before** we generate the spreadsheet, so that the spreadsheet is generated **after** the metadata has been added,

This is to allow the UI will show **Download Spreadsheet** instead of **Generate Spreadsheet** when a spreadsheet has already been generated.

**Comment by MightyAx** [comment]
for issue 1: stop naming the spreadsheet using the date time stamp,
using similar sequence to [triggering file revalidation](https://github.com/ebi-ait/hca-ebi-dev-team/tree/master/scripts/retrigger_file_validation)
remove the cloudUrl and checksum and set validation to file_not_uploaded

sending the file to s3 will overwrite the existing file and trigger the post back of the checksums, clearing the validation errors

doing this before we generate the spreadsheet will also fix item 3

**Comment by gabsie** [comment]
Just add the new emergent issues on this here, @MightyAx and we continue tracking this ticket.

**Comment by MightyAx** [comment]
Generating spreadsheet via Exporter redesign ideas:

1. Core: preempt the existing [`DefaultExporter.exportMetadata(ExportJob exportJob)`](https://github.com/ebi-ait/ingest-core/blob/9074e3af46032d50adbf06a984e8779ac978ef77/src/main/java/org/humancellatlas/ingest/exporter/DefaultExporter.java#L94)
    - This probably by renaming `exportMetadata()` -> `generateSpreadsheet()`
     - Put a message on a new spreadsheet generation queue, content is probably just the submission / export job
1. Exporter: new listener that generates the spreadsheet
    - hit the PATCH `/exportJobs/{id}/context` endpoint when you're done to add `"spreadsheetGeneration": "complete"` to the `exportJob.context`
1. in the [`ExportJobService`](https://github.com/ebi-ait/ingest-core/blob/9074e3af46032d50adbf06a984e8779ac978ef77/src/main/java/org/humancellatlas/ingest/export/job/ExportJobService.java#L57) you can trigger the original functionality if the patch sets the `spreadsheetGeneration` to "complete"

**Comment by Wkt8** [comment]
Amnon to catch-up with Alexie this morning. Future steps pending.

**Comment by MightyAx** [comment]
On the rework we have a draft PR in core and exporter with the general flow covered, but as we continue

Rather than having exporter call broker we should:

- [x] push the export code into client (its already mostly there as data collector)
- [x] Have the exporter run a version of [the aborted broker code](https://github.com/ebi-ait/ingest-broker/pull/44) for generating metadata
- [x] have the exporter trigger the new client spreadsheet exporter
- [x] send the generated metadata/spreadsheet to terra

**Comment by amnonkhen** [comment]
PR ebi-ait/ingest-exporter#46 implements step 4 in [Alexie's comment](https://app.zenhub.com/workspaces/dcp-ingest-product-development-5f71ca62a3cb47326bdc1b5c/issues/ebi-ait/dcp-ingest-central/845#issuecomment-1289169602)

The missing bits are:
- [x] integrate your function that creates the metadata json for the spreadsheet
- [ ] sort out the key name for the spreadsheet on terra.
These happen all on a new function I created: `SpreadsheetHandler.process_spreadsheet_metadata`

**Comment by MightyAx** [comment]
I've made the code "work" so far testing in the python console hasn't thrown any errors but I haven't yet sent any data to terra.
``` test.py
from unittest.mock import Mock
from exporter.terra.spreadsheet.exporter import SpreadsheetExporter
from exporter.terra.storage import TerraStorageClient
terra = Mock(spec=TerraStorageClient)
from exporter.ingest.service import IngestService
from hca_ingest.api.ingestapi import IngestApi
exporter = SpreadsheetExporter(IngestService(IngestApi('https://api.ingest.dev.archive.data.humancellatlas.org')), terra)
exporter.export_spreadsheet('','c578a42a-83d6-4123-b48c-dd726cd25c20','4f47ccdc-08b8-4176-8030-1e5e99c7f17f')
```

More manual testing and some unit tests required!

**Comment by ESapenaVentura** [comment]
Testing is next!

**Comment by ofanobilbao** [comment]
Amnon hoping to finish testing and fixing issues today

**Comment by amnonkhen** [comment]
code fixes & testing are in the following PRs:
- ebi-ait/ingest-client#60
- ebi-ait/ingest-exporter#45
- ebi-ait/ingest-core#147

see exporter test class: (see [tests/exporter/terra/spreadsheet/test_exporter.py](https://github.com/ebi-ait/ingest-exporter/blob/2a72b3961a5c583fdd0bf8b95ca5c0d1f88a253e/tests/exporter/terra/spreadsheet/test_exporter.py))

**Comment by MightyAx** [comment]
# Manual Testing
1. Initial Export
   - [x] Create project
   - [x] Export
   - [x] Verify spreadsheet
2. Updated Export For the same submission
   - [x] Create project
   - [x] Export
   - [x] Verify spreadsheet
   - [x] Update metadata (in same submission)
   - [x] Export
   - [x] Verify spreadsheet
3. Multi Submission Project Support (Expect Failure)
   - [x] Create project from small dataset
   - [x] Export
   - [x] Verify spreadsheet
   - [x] Create new submission from medium dataset in existing project
   - [x] Export
   - [x] Verify spreadsheet
4. Failed spreadsheet exports
   - [ ] A failed spreadsheet generation stops the export

**Comment by MightyAx** [comment]
# Integration (/System Tests)
- [x] Initial Export
- [x] Update Export for Same Submission
- [x] Multi Submission Project Support (Expect Failure)

**Comment by ESapenaVentura** [comment]
@MightyAx to set-up and start testing locally today

Maybe meet with @Mil-m to explain

**Comment by ESapenaVentura** [comment]
@MightyAx to set-up and start testing locally today

Maybe meet with @Mil-m to explain

**Comment by MightyAx** [comment]
ebi-ait/ingest-exporter#45 and ebi-ait/ingest-core#147 deployed to dev for testing.

**Comment by idazucchi** [comment]
alexie working on a bug - the exporter keeps all spreadsheets generated

**Comment by Wkt8** [comment]
A couple of changes still to be made to the flattener and then additional testing required. PR requires approvals.

**Comment by MightyAx** [comment]
We should ping Hannes and Trevor on slack once we have final version with examples on dev/prod.

**Comment by MightyAx** [comment]
The new ingest client is ready and I've tested it
@amnonkhen please review [the PR](https://github.com/ebi-ait/ingest-client/pull/61)

**Comment by MightyAx** [comment]
client pr merged and deployed to PyPI
@amnonkhen [Exporter PR](https://github.com/ebi-ait/ingest-exporter/pull/47) Ready for Review

**Comment by MightyAx** [comment]
PR Approved, One comment about logging has sparked my interest.
Just trying to finish that this morning so we can merge and manually test on dev today.

**Comment by MightyAx** [comment]
PR Approved, One comment about logging has sparked my interest.
Just trying to finish that this morning so we can merge and manually test on dev today.

@MightyAx to post to slack to invite group testing once merged to dev.

**Comment by MightyAx** [comment]
The latest ingest-broker, ingest-exporter (with ingest-client) and ingest-core have been deployed to dev for testing.

Initial tests show spreadsheets being sent to terra without the additional information from previous exports

**Comment by Mil-m** [comment]
the next steps:
1. Manual Tests - @MightyAx, @Mil-m 

- [x] Initial Export
- [x] Update metadata on same submission and re-export
- [x] Multi Submission Project Support (Expect Failure)

2. User Flow Test @ESapenaVentura

**Comment by ESapenaVentura** [comment]
Tried for user flow:

- Created a project and a submission, project with custom metadata (Manually filled), submission with `small_dataset.xlsx` testing spreadsheet https://dev.contribute.data.humancellatlas.org/submissions/detail?uuid=de4768d5-03a9-4aaf-8b92-ac0ef96af577&project=9b2a1653-b9bc-4b0e-befa-272784354a04
- Uploaded files, everything went correctly, and exported
- Once exported, I checked the staging area - The file was there! I checked the file metadata, the file descriptor and downloaded the file - it looks exactly like the project!

I am going to do another test, though - I am going to delete the file from the staging area and re-export - This should regenerate the file, with the same metadata/uuids/names

**Comment by ESapenaVentura** [comment]
Spreadsheet metadata has the exact same uuid and was regenerated successfully!

**Comment by ofanobilbao** [comment]
It has been tested multiple times, for different scenarios. We will move this to staging so that more people can do testing.

To merge the following need to be merged to master:

- ingest-broker,
- ingest-exporter (with ingest-client)
- ingest-core

**Comment by MightyAx** [comment]
deployments to staging were made successfully and testing began.

**Comment by MightyAx** [comment]
I’ve found an issue with spreadsheet exporting, I don’t think this effects us going to production, but it will need to fixed.
Once we’ve sent a spreadsheet (or most any metadata file actually) to terra, when we try to send an updated version it fails but gets marked as success.
I’ve posed this as an issue with metadata exporting before, and since we are using the same routines to send the metadata spreadsheet it fails in the same way.
So this bug effects any “edit and reexport” flow, but any initial export creates the project / submission just fine

**Comment by MightyAx** [comment]
For future reference: https://cloud.google.com/storage/docs/object-versioning

**Comment by MightyAx** [comment]
#### Issue to be made as a new ticket:
- [ ] **Updating an existing project that has already been exported will silently fail to send a new version of the metadata file.**
    - This is a wider problem with the metadata exporter, that now also affects metadata spreadsheet exporter
        - If a file exists it is assumed to be the correct version and no attempt is made to check/update it.
    - This is because file existence is used as the key to whether a file should be exported **at all**, a file that exists is assumed to be an in progress upload from another exporter pod, that this export should wait behind. When the exporter tries to figure out how long it should wait it is surprised and delighted that the file has already finished uploading and no further action is taken.
    - On some level this is necessitated by the "one export per experiment" metadata exporter design that is an out-of-date hold-over from DCPv1 which we no longer need to support.
    - We *could* move to a one export per submission design...

**Comment by Wkt8** [comment]
As above - new ticket at the design stage to be made - workaroundable. Doesn't block this ticket.

**Comment by MightyAx** [comment]
new ticket made: #873
ToDo: deploy to prod:

- [x] core
- [x] exporter
- [x] broker

**Comment by MightyAx** [comment]
Production Deployment and integration tests ran successfully

**Comment by MightyAx** [comment]
@MightyAx 
- [x] to check production for terra spreadsheets
@gabsie
- [ ] to notify downstream components by email

**Comment by MightyAx** [comment]
[This project](https://contribute.data.humancellatlas.org/submissions/detail?uuid=72705a81-c281-403d-96bd-fdb606f6afec&project=fc381e70-df1b-407d-8131-52ab523270bd) has been exported and generated this spreadsheet as expected.
[https://app.zenhub.com/files/299258791/14ba93bc-d204-4e8c-8df9-e315000cfa48/download](https://app.zenhub.com/files/299258791/14ba93bc-d204-4e8c-8df9-e315000cfa48/download)

**Comment by ESapenaVentura** [comment]
@gabsie to check the comment above with the checkbox ^^^

**Comment by Wkt8** [comment]
Still waiting on #873
